### PR TITLE
k8s.io/kms: drop direct dependency on klog

### PR DESCRIPTION
**Pull request by enj**
Signed-off-by: Monis Khan <mok@microsoft.com>

Builds on #120896

This will make #120225 drop `klog` altogether.

/kind cleanup
/assign aramase 

```release-note
NONE
```

### CODE REVIEW

**Review by aramase — COMMENTED** [review]
/lgtm

**Review by aramase — COMMENTED** [review]
/lgtm

**Comment by dims** [comment]
/assign @pohly 

@enj @aramase we don't yet have guidance from folks working on logging/klog about allowing `slog` in our code base

**Comment by dims** [comment]
also build failure:
```
vendor/k8s.io/kms/pkg/service/grpc_service.go:21:2: package log/slog is not in GOROOT (/usr/local/go/src/log/slog)
```

**Comment by enj** [comment]
> /assign @pohly
> 
> @enj @aramase we don't yet have guidance from folks working on logging/klog about allowing `slog` in our code base

FWIW the kms staging repo is for external consumption, hence I was trying to limit its dep tree.

> also build failure:
> 
> ```
> vendor/k8s.io/kms/pkg/service/grpc_service.go:21:2: package log/slog is not in GOROOT (/usr/local/go/src/log/slog)
> ```

I figured it would fail if we didn't support go1.21 yet :)

**Comment by pohly** [comment]
Using `slog.Debug` means that logging will use the global default slog handler. That's okay(is), that handler can be set by a binary such that the output goes to klog, if that is what the binary is using. See:
- https://github.com/go-logr/logr/pull/222
- https://github.com/kubernetes/kubernetes/pull/120696

However, it means that the code doesn't support contextual logging. My proposal is that we continue to use klog for handling access to a logger and the go-logr API for log calls and not switch to slog:
- https://github.com/kubernetes/enhancements/pull/4219

How important is it to remove the klog dependency? This is doable, but requires more changes than planned right now. For example, there are helpers in klog like `klog.KObj` which are needed for structured logging. We would need a separate package with all the calls that we intend to keep using.

**Comment by pohly** [comment]
For the sake of simplicity, that other package could be `klog/v3`... then all it takes in code with full conversion to structured, contextual logging would be a `s;klog/v2;klog/v3`. But a bit more thought will be needed regarding interoperability of the two versions...

**Comment by enj** [comment]
> How important is it to remove the klog dependency?

Quite high for the kms staging repo because the code in there has no use in k/k itself, it is just a thin wrapper around the proto APIs to make it easy for someone to build a plugin.  The reason for using slog is so that a plugin can easily override the logger and doesn't have to deal with any of the weirdness associated with klog.  I can certainly make this code just use the log package instead if we really want to.  Alternatively I can delete the log statements here and move them into the mock implementation.

**Comment by enj** [comment]
Note also that this code doesn't need contextual logging because it is never going to be used in-tree.  Basically this change should have no bearing on what we decide to do as a whole for the Kube project (where I agree using klog is the correct choice).

**Comment by pohly** [comment]
If you want to get rid of klog, then I'd prefer to have no log calls in that package at all. That's the only short-term solution anyway, or are you ready to make Go 1.21 the required version for it?

**Comment by pohly** [comment]
> Note also that this code doesn't need contextual logging because it is never going to be used in-tree. 

Who says that contextual logging is only something that is done in-tree? :smile:

The concept also applies out-of-tree. For example, Cluster API is using it and the controller-runtime has support for it.

**Comment by enj** [comment]
/hold cancel

I removed all usage of `slog` from this PR.

File: staging/src/k8s.io/kms/internal/plugins/_mock/Dockerfile [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM golang:1.20.1-bullseye as builder
+FROM golang:1.21.1-bullseye as builder
```
Comment (aramase) [inline staging/src/k8s.io/kms/internal/plugins/_mock/Dockerfile:15]:
```suggestion
FROM golang:1.21.3-bullseye as builder
```

**Review by aramase — COMMENTED** [review]
/lgtm

**Pull request merged**